client/cli/src/commands/utils.rs

```diff
@@ -0,0 +1,279 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! subcommand utilities
+use std::{io::Read, path::PathBuf};
+use sp_core::{Pair, Public, crypto::{Ss58Codec,Derive, Ss58AddressFormat}, hexdisplay::HexDisplay};
+use sp_runtime::{
+	MultiSignature, MultiSigner, generic::{UncheckedExtrinsic, SignedPayload},
+	traits::IdentifyAccount, AccountId32,
+};
+use parity_scale_codec::{Encode, Decode};
+use serde_json::json;
+use sp_runtime::traits::{
+	SignedExtension, StaticLookup,
+};
+use crate::{error::{self, Error}, SharedParams, arg_enums::OutputType};
+
+/// Public key type for Runtime
+pub type PublicFor<R> = <<R as RuntimeAdapter>::Pair as Pair>::Public;
+/// Seed type for Runtime
+pub type SeedFor<R> = <<R as RuntimeAdapter>::Pair as Pair>::Seed;
+/// AccountIndex type for Runtime
+pub type IndexFor<R> = <<R as RuntimeAdapter>::Runtime as frame_system::Trait>::Index;
+/// AccountId type for Runtime
+pub type AccountIdFor<R> = <<R as RuntimeAdapter>::Runtime as frame_system::Trait>::AccountId;
+/// Balance type
+pub type BalanceFor<R> = <<R as RuntimeAdapter>::Runtime as pallet_balances::Trait>::Balance;
+/// Call type for Runtime
+pub type CallFor<R> = <<R as RuntimeAdapter>::Runtime as frame_system::Trait>::Call;
+/// Address type for runtime.
+pub type AddressFor<R> = <<<R as RuntimeAdapter>::Runtime as frame_system::Trait>::Lookup as StaticLookup>::Source;
+/// Hash for runtime.
+pub type HashFor<R> = <<R as RuntimeAdapter>::Runtime as frame_system::Trait>::Hash;
+
+/// Runtime adapter for signing utilities
+pub trait RuntimeAdapter: Sized {
```

Comment:
Hah! Funny, I'm also implementing something similar to this

Reply:
:thinking: I think this is great actually. It means the API of the CLI is getting correct.

Reply:
I wonder if this is correct though... @seunlanlege does it make sense that it is in the CLI? It looks like parameters you want for a blockchain. Maybe it would make more sense to have this in sc_service?

Reply:
> It looks like parameters you want for a blockchain.

Hmm, not sure what you mean, the core param here is the `Runtime` and crypto types.

> Maybe it would make more sense to have this in sc_service.

You're referring to the `RuntimeAdapter` trait, right? I'm not sure if it fits in sc_service since this is primarily used for cli, or maybe i don't fully understand where you're coming from.  I'm actually considering moving it to `sc-client-api`

Reply:
> Hah! Funny, I'm also implementing something similar to this.

cool! what for though?

Reply:
In sc-cli you should have things more related to building a CLI with substrate. While in sc-service you have things more related to gluing up your blockchain in general (runtime, network etc...) (@tomaka maybe correct me if I'm wrong).

> cool! what for though?

I'm currently implementing a trait in sc-cli to replace the VersionInfo and the `load_spec` function of sc-cli. A CLI binary will have to impl this new trait on their structopt and it will "enrich"/"augment" the structopt to have all the stuff needed for a substrate CLI (mostly version information, program name, authors, etc... and load_spec).

Reply:
> we might want to consider RuntimeAdapter something to be implemented directly on Runtime. This is very much runtime dependent, and looks a lot like the TransacationSubmitter 

I'm closing this conversation in favor of https://github.com/paritytech/substrate/pull/4954#discussion_r394155651